### PR TITLE
[2/n] lora - Shared outer experts and support qwen3_30b_a3b_instruct

### PR DESCRIPTION
**Pull request by yushengsu-thu**
## Motivation

Generated_image

- Support: hared outer expert LoRA

- Support:  Qwen3-30B-A3B-Instruct-2507

## Modifications

- Shared outer expert LoRA: Supports MoE adapters where gate_up lora_A and down lora_B are shared across all experts (expert_dim=1) instead of per-expert. The memory pool allocates collapsed buffers; at runtime the shared weights are expand()-ed to num_experts in the Triton MoE runner. Auto-detected from adapter weights or forced via --experts-shared-outer-loras.

- Qwen3-30B-A3B-Instruct-2507 compatibility: Adds _rename_expert_w_to_proj to normalize w1/w3/w2 naming to gate_proj/up_proj/down_proj; generalizes num_experts detection (num_local_experts, n_routed_experts); handles get_text_config() for VL models; fixes N-dimension masking in the Triton sgemm_lora_b kernel.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/lora/layers.py [hunk 1]
```diff
@@ -711,6 +711,9 @@ def __init__(
         # initializes FusedMoE with its own moe_runner for base path
         super().__init__(base_layer, lora_backend)
 
+        self.experts_shared_outer_loras: bool = False
+        self.quant_method = base_layer.quant_method
```
Comment (sshleifer) [inline python/sglang/srt/lora/layers.py:715]: would comment that this doesnt actually quantize the loras, it confused me at first.

File: python/sglang/srt/lora/layers.py [hunk 2]
```diff
@@ -838,35 +842,75 @@ def slice_lora_a_weights(self, A: torch.Tensor, tp_rank: int):
...
           down_proj_moe A:    [rank, intermediate_size] — input is sharded intermediate
         """
         if self.tp_size <= 1:
             return A
+        if isinstance(A, dict):
+            return {
+                eid: self._slice_moe_a_2d(w, tp_rank, target_module)
+                for eid, w in A.items()
+            }
+        if isinstance(A, torch.Tensor) and A.dim() == 3:
```
Comment (sshleifer) [inline python/sglang/srt/lora/layers.py]: we could have a fast path

elif target_module != "down_proj_moe": return A

to avoid the no-op stack loop (If I understand correctly)

File: python/sglang/srt/lora/layers.py [hunk 3]
```diff
@@ -838,35 +842,75 @@ def slice_lora_a_weights(self, A: torch.Tensor, tp_rank: int):
...
+        return self._slice_moe_a_2d(A, tp_rank, target_module)
+
+    def _slice_moe_a_2d(
+        self, A: torch.Tensor, tp_rank: int, target_module: str
+    ) -> torch.Tensor:
         if target_module == "down_proj_moe":
             shard_size = self.intermediate_size_per_partition
             start = tp_rank * shard_size
             end = start + shard_size
             return A[:, start:end].contiguous()
```
Comment (sshleifer) [inline python/sglang/srt/lora/layers.py]: you can always slice the trailing dimension with

```return A[..., start:end].contiguous()```

then don't need stack loop

File: python/sglang/srt/lora/layers.py [hunk 4]
```diff
@@ -838,35 +842,75 @@ def slice_lora_a_weights(self, A: torch.Tensor, tp_rank: int):
...
+        if isinstance(B, torch.Tensor) and B.dim() == 3:
+            return torch.stack(
+                [
+                    self._slice_moe_b_2d(B[i], tp_rank, target_module)
+                    for i in range(B.shape[0])
+                ]
+            )
+        return self._slice_moe_b_2d(B, tp_rank, target_module)
+
+    def _slice_moe_b_2d(
```
Comment (sshleifer) [inline python/sglang/srt/lora/layers.py:916]: 1. can avoid some stack
2. its not clear to me that the kernel likes this format

File: python/sglang/srt/lora/lora.py [hunk 5]
```diff
@@ -216,7 +236,10 @@ def normalize_gate_up_proj(
                 # If gate_up_proj is already stacked, we normalize it following the SGL convention
                 gate_up_name = weight_name
                 if "lora_A" in weight_name:
-                    weights[gate_up_name] = weights[gate_up_name].repeat(2, 1)
+                    ndim = weights[gate_up_name].dim()
+                    repeat_dims = [1] * ndim
+                    repeat_dims[ndim - 2] = 2
+                    weights[gate_up_name] = weights[gate_up_name].repeat(*repeat_dims)
```
Comment (sshleifer) [inline python/sglang/srt/lora/lora.py:242]: What is this for?
(I also dont understand the logic on main)

Feel free to ignore!

File: python/sglang/srt/lora/lora_manager.py [hunk 6]
```diff
@@ -411,6 +434,26 @@ def init_lora_adapters(self, lora_paths: Optional[List[LoRARef]] = None):
                         f"Failed to load LoRA adapter {lora_ref.lora_name}: {result.error_message}"
                     )
 
+    def _detect_shared_outer_loras(self) -> bool:
```
Comment (sshleifer) [inline python/sglang/srt/lora/lora_manager.py:437]: Do we call this every time we load/unload loras on same server?
If so, we should cache the result because it wont change during the server lifetime

File: python/sglang/srt/lora/lora_manager.py [hunk 7]
```diff
@@ -424,24 +467,27 @@ def init_lora_shapes(
...
+                        # Resolve by scanning the base model for all
+                        # LoRA-compatible linear modules.
+                        adapter_target_modules = auto_detect_lora_target_modules(
+                            self.base_model
+                        )
+                        logger.info(
+                            "LoRA adapter '%s' uses target_modules='%s'. "
+                            "Resolved to %s by inspecting the base model.",
+                            self.lora_refs[lora_id].lora_name,
+                            config.target_modules,
```
Comment (sshleifer) [inline python/sglang/srt/lora/lora_manager.py:486]: will this print a superset of the actual modules? 
cause of
`normalized & _KNOWN_LORA_TARGET_MODULES`

File: python/sglang/srt/lora/lora_manager.py [hunk 8]
```diff
@@ -672,16 +719,17 @@ def init_lora_modules(self):
             # The module should be converted if it is included in target_names
             if module_name.split(".")[-1] in self.target_modules:
```
Comment (sshleifer) [inline python/sglang/srt/lora/lora_manager.py:720]: lm_head handled elsewhere?

File: python/sglang/srt/lora/lora_moe_runners.py [hunk 9]
```diff
@@ -469,16 +475,11 @@ def _add_lora_gate_up_delta(
 
         r = lora_info.max_lora_rank
         gate_up_a = lora_info.gate_up_lora_a_weights
+        if lora_info.experts_shared_outer_loras:
+            gate_up_a = gate_up_a.expand(-1, lora_info.num_experts, -1, -1)
```
Comment (sshleifer) [inline python/sglang/srt/lora/lora_moe_runners.py:479]: link 4 later. This seems suboptimal perf wise

File: python/sglang/srt/lora/mem_pool.py [hunk 10]
```diff
@@ -587,41 +613,77 @@ def load_lora_weight_tensor(
...
+                                buffer_id, 0, : lora_rank * c, :
+                            ]
+                            load_lora_weight_tensor(buffer_view, weights[0])
+                        elif isinstance(weights, dict) and len(weights) > 0:
+                            rep = next(iter(weights.values()))
+                            buffer_view = target_buffer[
+                                buffer_id, 0, : lora_rank * c, :
+                            ]
+                            load_lora_weight_tensor(buffer_view, rep)
+                        else:
```
Comment (sshleifer) [inline python/sglang/srt/lora/mem_pool.py:633]: what case is this? raise?

File: test/registered/lora/test_lora_qwen3_30b_a3b_instruct_2507_logprob_diff.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+LORA_BACKEND = "triton"
+MAX_LORA_RANK = 32
+TP_SIZE = 8
+DISABLE_CUDA_GRAPH = True
+MOE_RUNNER_BACKEND = "triton"
+EXPERTS_SHARED_OUTER_LORAS = True
+PREFILL_ATTENTION_BACKEND = "fa4"
+DECODE_ATTENTION_BACKEND = "fa4"
+
+KL_THRESHOLD = 1e-2
```
Comment (sshleifer) [inline test/registered/lora/test_lora_qwen3_30b_a3b_instruct_2507_logprob_diff.py]: I think we can set this lower

File: test/registered/lora/test_lora_qwen3_8b_logprob_diff.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+BASE_MODEL = "Qwen/Qwen3-8B"
+LORA_HF_REPO = "yushengsu/lora-diff-Qwen3-8B"
+LORA_BACKEND = "triton"
+MAX_LORA_RANK = 32
+TP_SIZE = 1
+DISABLE_CUDA_GRAPH = True
+PREFILL_ATTENTION_BACKEND = "fa4"
+DECODE_ATTENTION_BACKEND = "fa4"
+
+KL_THRESHOLD = 1e-2
```
Comment (sshleifer) [inline test/registered/lora/test_lora_qwen3_8b_logprob_diff.py:54]: lower

**Review by sshleifer — APPROVED** [review]
Thans!

File: python/sglang/srt/lora/lora.py (hunk 5 above)
Comment (yushengsu-thu) [reply python/sglang/srt/lora/lora.py:242]: will need to refactor here

File: python/sglang/srt/lora/lora_manager.py (hunk 6 above)
Comment (yushengsu-thu) [reply python/sglang/srt/lora/lora_manager.py:437]: need to check after /n series PRs merged

File: python/sglang/srt/lora/layers.py (hunk 3 above)
Comment (yushengsu-thu) [reply python/sglang/srt/lora/layers.py]: done

File: python/sglang/srt/lora/lora_manager.py (hunk 8 above)
Comment (yushengsu-thu) [reply python/sglang/srt/lora/lora_manager.py:720]: yes

File: python/sglang/srt/lora/mem_pool.py (hunk 10 above)
Comment (yushengsu-thu) [reply python/sglang/srt/lora/mem_pool.py:633]: done

File: python/sglang/srt/lora/lora_moe_runners.py (hunk 9 above)
Comment (Fridge003) [reply python/sglang/srt/lora/lora_moe_runners.py:479]: Rather than expanding, we might create a new kernel that points all the expert loras into the same position

**Review by Fridge003 — APPROVED** [review]
